// @okta/okta-angular/esm2015/src/okta/components/callback.component.js
/*
 * Copyright (c) 2017-Present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */
import * as tslib_1 from "tslib";
import { Component, Optional, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { OktaAuthService } from '../services/okta.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
let OktaCallbackComponent = class OktaCallbackComponent {
    constructor(okta, router, injector) {
        this.okta = okta;
        this.router = router;
        this.injector = injector;
    }
    ngOnInit() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                // Parse code or tokens from the URL, store tokens in the TokenManager, and redirect back to the originalUri
                yield this.okta.handleLoginRedirect();
            }
            catch (e) {
                // Callback from social IDP. Show custom login page to continue.
                if (this.okta.isInteractionRequiredError(e) && this.injector) {
                    const { onAuthResume, onAuthRequired } = this.okta.getOktaConfig();
                    const callbackFn = onAuthResume || onAuthRequired;
                    if (callbackFn) {
                        callbackFn(this.okta, this.injector);
                        return;
                    }
                }
                this.error = e.toString();
            }
        });
    }
};
OktaCallbackComponent.ɵfac = function OktaCallbackComponent_Factory(t) { return new (t || OktaCallbackComponent)(ɵngcc0.ɵɵdirectiveInject(OktaAuthService), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector, 8)); };
OktaCallbackComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: OktaCallbackComponent, selectors: [["ng-component"]], decls: 2, vars: 1, template: function OktaCallbackComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate(ctx.error);
    } }, encapsulation: 2 });
OktaCallbackComponent = tslib_1.__decorate([ tslib_1.__param(2, Optional()),
    tslib_1.__metadata("design:paramtypes", [OktaAuthService, Router, Injector])
], OktaCallbackComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(OktaCallbackComponent, [{
        type: Component,
        args: [{
                template: `<div>{{error}}</div>`
            }]
    }], function () { return [{ type: OktaAuthService }, { type: ɵngcc1.Router }, { type: ɵngcc0.Injector, decorators: [{
                type: Optional
            }] }]; }, null); })();
export { OktaCallbackComponent };
